GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs

```diff
@@ -37,91 +37,85 @@ partial class RepoObjectsTree : IMenuItemFactory
         /// </summary>
         private MenuItemsGenerator<TagNode> _tagNodeMenuItems;
 
-        private void ContextMenuAddExpandCollapseTree(ContextMenuStrip contextMenu)
+        /// <summary>Toggles the <paramref name="items"/> <see cref="ToolStripItem.Visible"/>
+        /// as well as <see cref="ToolStripItem.Enabled"/> properties depending on <paramref name="enabled"/>.
+        /// Prefer this over only toggling the visibility of an item to enable determining whether the context menu will (once open)
+        /// contain any visible items via <see cref="ToolStripItem.Enabled"/> in <see cref="contextMenu_Opening(object, CancelEventArgs)"/>
+        /// even before the menu itself (as the visual parent) is visble and <see cref="ToolStripItem.Visible"/> of any item therefore false.</summary>
+        private void ToggleMenuItems(bool enabled, params ToolStripItem[] items) => items.ForEach(item => item.Visible = item.Enabled = enabled);
+
+        private void ToggleMenuItems<TNode>(MenuItemsGenerator<TNode> generator, Func<ToolStripItemWithKey, bool> isEnabled) where TNode : class, INode
+            => generator.ForEach(i => ToggleMenuItems(isEnabled(i), i.Item));
+
+        /* add Expand All / Collapse All menu entry
+         * depending on whether node is expanded or collapsed and has child nodes at all */
+        private void ToggleExpandCollapseContextMenu(ContextMenuStrip contextMenu, NodeBase[] selectedNodes)
```

Comment:
The non-script menu items should be added in the designer.
Just their `Visible` and `Enabled` state should be set in dependency on the `selectedNodes`.

`Filter for selected` belongs to the section of branch / tag actions IMO.
Scripts should be appended as last items after a separator - same as in the RevisionGrid.

Reply:
> `Filter for selected` belongs to the section of branch / tag actions IMO.
> Scripts should be appended as last items after a separator - same as in the RevisionGrid.

@mstv I understand that this is in answer of my question (3.2).

> The non-script menu items should be added in the designer.
> Just their `Visible` and `Enabled` state should be set in dependency on the `selectedNodes`.

I don't understand this. Please rephrase?

Reply:
All fixed menu items, i.e. all but those dynamically created for scripts, should be added in the `.Designer.cs` and remain in `Items` all the time.
Whether they are shown to the user, is controlled using `Enabled` and `Visible`.
Then there will be no need for `AddOnce()` etc.

Reply:
@mstv I've replaced the `AddOnce()` with a `SetLastItem()` method in the latest version, but notice the pattern hasn't changed. This method is used in it for menu item instances that are shared between multiple context menus (sort, expand/collapse, Filter for selected). It's necessary to add those items to the menu over and over again because adding them to a different menu removes them from this one.

For example: The `Filter for selected` item is shared between the local branch menu **L** and the remote branch menu **R**. When opening **L**, the item is added to **L**. When opening **R**, the item is added to **R** - which implicitly removes it from **L**, because an item can only be part of one context menu. That's why it has to be added to **L** again when re-opening **L**.

In order to get rid of this pattern, I'd have to rewrite the context menu to prevent sharing menu item instances, e.g. by adding clones instead of the same instance. That should prevent the accidental removal of the item from other context menus and make re-adding unnecessary. Do you want me to do that?